### PR TITLE
Refactors Bindinds to use Async Workers

### PR DESCRIPTION
**Pull request by daniel-j-h**
This fixes concurrency issues reported in #189.

The code changed quite dramatically, better take a look at the files directly
- https://github.com/Project-OSRM/node-osrm/blob/async_worker/src/node_osrm.hpp
- https://github.com/Project-OSRM/node-osrm/blob/async_worker/src/node_osrm.cpp
- https://github.com/Project-OSRM/node-osrm/blob/async_worker/src/node_osrm_support.hpp

References:
- https://github.com/nodejs/nan#asynchronous-work-helpers
- https://github.com/nodejs/nan/blob/master/doc/asyncworker.md#api_nan_async_worker
- https://github.com/nodejs/nan/blob/master/doc/asyncworker.md#api_nan_async_queue_worker

/cc @TheMarex rebased into the initial commit you made for me

### CODE REVIEW

**Comment by daniel-j-h** [comment]
The support code could benefit from some refactoring.
I didn't touch it except for moving it into a separate header.

File: src/node_osrm.cpp [hunk 1]
```diff
@@ -1,1161 +1,173 @@
...
-
-    if (path->IsUndefined() && !engine_config->use_shared_memory)
-    {
-        Nan::ThrowError("Shared_memory must be enabled if no path is "
-                        "specified");
-        return engine_config_ptr();
-    }
-
-    return engine_config;
+    static Nan::Persistent<v8::Function> init;
```
Comment (daniel-j-h) [inline src/node_osrm.cpp]: This function is needed for the thread-safe singleton impl. without relying on global constructors.

File: test/index.js [hunk 2]
```diff
@@ -2,12 +2,6 @@ var OSRM = require('../');
 var test = require('tape');
 var berlin_path = require('./osrm-data-path').data_path;
 
-test('constructor: throws if new keyword is not used', function(assert) {
```
Comment (TheMarex) [inline test/index.js]: Why did you remove this test?

File: src/node_osrm.cpp [hunk 3]
```diff
@@ -1,1161 +1,173 @@
...
-
-        if (!bearings->IsArray())
+        try
         {
-            Nan::ThrowError("Bearings must be an array of arrays of numbers");
-            return false;
-        }
+            auto config = argumentsToEngineConfig(info);
+            if (!config)
+                return;
```
Comment (daniel-j-h) [inline src/node_osrm.cpp]: Why do we simply return here? (Was already in the old code like that)
Shouldn't this throw an error? This is done multiple times if I recall correctly.

File: src/node_osrm.cpp [hunk 4]
```diff
@@ -1,1161 +1,173 @@
...
 
-        v8::Local<v8::Array> timestamps_array = v8::Local<v8::Array>::Cast(timestamps);
-
-        if (params->coordinates.size() != timestamps_array->Length())
+        void Execute() override try
         {
-            Nan::ThrowError("Timestamp array must have the same size as the coordinates "
-                            "array");
-            return match_parameters_ptr();
+            const auto status = ((*osrm).*(service))(*params, result);
```
Comment (daniel-j-h) [inline src/node_osrm.cpp]: Calling through a mem fn ptr, lifted to smart pointers.

File: src/node_osrm.cpp [hunk 5]
```diff
@@ -1,1161 +1,173 @@
...
-    }
-    if (try_catch.HasCaught())
-    {
-        Nan::FatalException(try_catch);
-    }
-    closure->machine->Unref();
-    closure->cb.Reset();
-    delete closure;
+    auto *callback = new Nan::Callback{info[info.Length() - 1].As<v8::Function>()};
+    Nan::AsyncQueueWorker(new Worker{self->this_, std::move(params), service, callback});
```
Comment (daniel-j-h) [inline src/node_osrm.cpp]: `NAN` takes ownership and `delete`s.

File: test/index.js (hunk 2 above)
Comment (daniel-j-h) [inline test/index.js]: Since there is no reason not to support non-construct calls. So Instead of throwing I added this, take a look:
https://github.com/Project-OSRM/node-osrm/blob/ea36088732c7b3100ffea1192ca3978c10c767bd/src/node_osrm.cpp#L54-L76

Comment (TheMarex) [inline test/index.js]: This is to capture errors for accidental instantiation. Imagine `var OSRM = require('osrm')();` instead of `var OSRM = require('osrm');`

Comment (daniel-j-h) [inline test/index.js]: Okay, will revoke this change.

**Comment by springmeyer** [comment]
@daniel-j-h - can you share more about what was wrong with the original code, what the specific fix needed was, and how you confirmed the fix impacted performance? I'm asking because its not clear to me whether the fixes here apply or are relevant to other async code I maintain for other node c++ modules.

**Comment by daniel-j-h** [comment]
@springmeyer a couple of weeks before this pull request I learned about npm / node / nan by writing node bindings for @danpat's [route-annotator](https://github.com/mapbox/route-annotator/tree/d4aa6ea7631accc595951e9c37cb30aecbdca5ea/src). For making its features asynchronous I used [NAN's beautiful async. abstractions](https://github.com/nodejs/nan/tree/62e7c84b62361f5f22a367504e741cdaf114edbe#asynchronous-work-helpers). More concretely: 1/ inherit from `Nan::AsyncWorker` implementing the interface, 2/ enqueue the worker with `Nan::AsyncQueueWorker`.

If you take a look at the code before this changeset, `node-osrm` used `libuv` directly as in `uv_work_t`, `uv_queue_work` etc. --- when the performance regression came up during the sprint, I quickly prototyped the switch to NAN's async. abstractions for the Match service for which @TheMarex and I verified that it solved the problem. Eventually I rewrote all services (and added some abstractions that allows us to share code between the services).

Looping in libuv-wizard @kkaefer here.

**Comment by springmeyer** [comment]
@daniel-j-h - Thank you for the clarification. If I am understanding correctly then:

   - The code using `uv_work_t/uv_queue_work` directly was buggy somehow and not performing well
   - Porting to Nan fixed the issue
   - You benchmarked to confirm the performance difference

Is that correct? If so, there are dozens of node c++ modules that we (Mapbox) maintain which need fixed. So my further questions are:

  - Was the performance problem restricted to node v4 and not node v0.10?
  - Do you understand what Nan is doing that helped perf?

If you don't know no problem, I'm just trying to understand before diving in to figure out whether we need to change many node c++ modules and why/how. Thanks!

**Comment by daniel-j-h** [comment]
> If I am understanding correctly then:
> 
> - The code using uv_work_t/uv_queue_work directly was buggy somehow and not performing well
> - Porting to Nan fixed the issue
> - You benchmarked to confirm the performance difference
> 
> Is that correct?

The code using `uv_work_t` / `uv_queue_work` worked fine until we switched to Node 4.x (which updated the bundled `libuv` version, too). We never traced down the changes in `libuv` or the reason for the limited concurrency beyond the direct `libuv` usage, because the `AsyncWorker` approach fixed it and refactored the code at the same time quite nicely.

If you want to dig into this, this is what I would do:
- checkout the commit before this changeset
- get familiar with the source code that is using `libuv` a bit (maybe you find issues in it)
- reproduce concurrency issue (see the original issue at the top)
- read the Nodejs Changelog and get the `libuv` version number that comes bundled with it
- check the `libuv` docs and source for this version and the features the source code is using
- build a small self contained example reproducing the issue

>  Was the performance problem restricted to node v4 and not node v0.10?

Yes, I'm not sure though what else changed since I haven't had a close eye on `node-osrm` before, but I think that was the reason. Maybe @TheMarex or @lbud know more about this.

>  Do you understand what Nan is doing that helped perf? Is it using the libuv API differently?

From reading the NAN source, it looks like a really thin wrapper around `libuv`'s `uv_work_t` / `uv_queue_work`. Take a look here:

- [AsyncWorker](https://github.com/nodejs/nan/blob/62e7c84b62361f5f22a367504e741cdaf114edbe/nan.h#L1484-L1590)
- [AsyncQueueWorker](https://github.com/nodejs/nan/blob/62e7c84b62361f5f22a367504e741cdaf114edbe/nan.h#L1690-L1708)

Hope that helps.

**Comment by springmeyer** [comment]
Thanks @daniel-j-h - I'll try to replicate and dig in when I have time.

**Comment by danpat** [comment]
@springmeyer FWIW, it may be that the way `node-osrm` was using `uv_work_queue` was incorrect, and other usages @Mapbox are just fine.

We noticed this when we were running benchmarks - we were maxing out at 2 CPUs in use and couldn't drive usage any higher.  If you aren't seeing that performance problem elsewhere, then it's probably a non-issue.

Rather than diving into what was wrong with our direct usage of `uv_work_queue`, @daniel-j-h just swapped in `AsyncWorker` and we got our performance back.  This tidied up the code and simplified things a lot, so we ran with it.
